File: apollo/src/main/java/io/muun/apollo/domain/model/FeeWindow.java
package io.muun.apollo.domain.model;

import io.muun.common.Rules;
import io.muun.common.utils.Preconditions;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.threeten.bp.ZonedDateTime;

import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;
import javax.validation.constraints.NotNull;

public class FeeWindow {

    // time before this fee estimation is considered too old to use
    private static final int EXPIRATION_TIME_MINUTES = 5;

    @NotNull
    public final Long houstonId;

    @NotNull
    public final ZonedDateTime fetchDate;

    @NotNull
    public final SortedMap<Integer, Double> targetedFees = new TreeMap<>();

    @NotNull
    public final Integer fastConfTarget;

    @NotNull
    public final Integer mediumConfTarget;

    @NotNull
    public final Integer slowConfTarget;

    /**
     * Constructor.
     */
    public FeeWindow(@NotNull Long houstonId,
                     @NotNull ZonedDateTime fetchDate,
                     @NotNull Map<Integer, Double> targetedFees,
                     @NotNull Integer fastConfTarget,
                     @NotNull Integer mediumConfTarget,
                     @NotNull Integer slowConfTarget) {

        this.houstonId = houstonId;
        this.fetchDate = fetchDate;
        this.targetedFees.putAll(targetedFees);
        this.fastConfTarget = fastConfTarget;
        this.mediumConfTarget = mediumConfTarget;
        this.slowConfTarget = slowConfTarget;
    }

    /**
     * Return true if this FeeWindow is recent enough to be used.
     */
    @JsonIgnore
    public boolean isRecent() {
        return ZonedDateTime
                .now(fetchDate.getZone())
                .minusMinutes(EXPIRATION_TIME_MINUTES)
                .isBefore(fetchDate);
    }

    /**
     * Get the fastest fee rate, in satoshis per byte.
     */
    public double getFastestFeeInSatoshisPerByte() {
        final int shortestTarget = targetedFees.firstKey();
        final double satoshisPerByte = targetedFees.get(shortestTarget);

        return satoshisPerByte;
    }

    /**
     * Get the minimum available fee rate that will hit a given confirmation target. We make no
     * guesses (no averages or interpolations), so we might overshoot the fee if data is too sparse.
     */
    public double getMinimumFeeInSatoshisPerByte(int confirmationTarget) {
        Preconditions.checkPositive(confirmationTarget);

        // Walk the available targets backwards, finding the highest target below the given one:
        for (int closestTarget = confirmationTarget; closestTarget > 0; closestTarget--) {
            if (targetedFees.containsKey(closestTarget)) {
                // Found! This is the lowest fee rate that hits the given target.
                return targetedFees.get(closestTarget);
            }
        }

        // No result? This is odd, but not illogical. It means *all* of our available targets
        // are above the requested one. Let's use the fastest:
        final int lowestTarget = targetedFees.firstKey();
        return targetedFees.get(lowestTarget);
    }

    /**
     * Migrate FeeWindow to start using dynamic fee targets, set by houston. We'll initialize with
     * previous fixed values.
     */
    public FeeWindow initDynamicFeeTargets() {
        return new FeeWindow(
                houstonId,
                fetchDate,
                targetedFees,
                Rules.CONF_TARGET_FAST,
                Rules.CONF_TARGET_MID,
                Rules.CONF_TARGET_SLOW
        );
    }
}

File: libwallet/operation/fee_window.go
package operation

import (
	"fmt"
	"math"
)

const swapV2ConfTarget = 250 // Approx 2 days

type FeeWindow struct {
	TargetedFees map[uint]float64
}

// SwapFeeRate gets the appropriate fee rate for a given swap (depends on confirmations needed).
// Useful method for when swap doesn't have a fixed amount (e.g AmountLessInvoices + use all funds).
func (f *FeeWindow) SwapFeeRate(confirmationsNeeded uint) (float64, error) {
	if confirmationsNeeded == 0 {
		return f.MinimumFeeRate(swapV2ConfTarget)
	}

	return f.FastestFeeRate(), nil
}

// MinimumFeeRate gets the minimum available fee rate that will hit a given confirmation target. We
// make no guesses (no averages or interpolations), so we might overshoot the fee if data is too
// sparse.
// Note: the lower the confirmation target, the faster the tx will confirm, and greater the
// fee(rate) will be.
func (f *FeeWindow) MinimumFeeRate(confirmationTarget uint) (float64, error) {

	if confirmationTarget <= 0 {
		return 0, fmt.Errorf("can't get feeRate. Expected positive confirmation target, got %v", confirmationTarget)
	}

	// Walk the available targets backwards, finding the highest target below the given one:
	for closestTarget := confirmationTarget; closestTarget > 0; closestTarget-- {
		if feeRate, containsKey := f.TargetedFees[closestTarget]; containsKey {
			// Found! This is the lowest fee rate that hits the given target.
			return feeRate, nil
		}
	}

	// No result? This is odd, but not illogical. It means *all* of our available targets
	// are above the requested one. Let's use the fastest:
	return f.FastestFeeRate(), nil
}

// FastestFeeRate gets the fastest fee rate, in satoshis per weight unit.
func (f *FeeWindow) FastestFeeRate() float64 {

	var lowestTarget uint = math.MaxUint32
	for k := range f.TargetedFees {
		if k < lowestTarget {
			lowestTarget = k
		}
	}

	return f.TargetedFees[lowestTarget]
}

// NextHighestBlock finds the next highest confirmation/block target for a certain feeRate. Let me
// explain, we have a map that associates a conf-target with a fee rate. Now we want to know
// associate a conf-target with a given fee rate. We want the NEXT conf-target as we usually want
// this data for predictions or estimations and this makes the predictions for the fee rate to
// "fall on the correct side" (e.g when estimating max time to confirmation for a given fee rate).
// Note: code is not our best work of art. The target < next comparison is to account for our
// TargetedFees map not necessarily being sorted.
func (f *FeeWindow) NextHighestBlock(feeRate float64) uint {
	next := uint(math.MaxUint32)
	for target, rate := range f.TargetedFees {
		if rate <= feeRate && target < next {
			next = target
		}
	}
	if next == math.MaxUint32 {
		return 0 // 0 is a not valid targetBlock, we use it to signal target not found
	}
	return next
}
